parquet-hadoop/src/main/java/parquet/hadoop/util/CompatibilityUtil.java

```diff
@@ -0,0 +1,398 @@
+ * limitations under the License.
+ */
+package parquet.hadoop.util;
+
+import org.apache.hadoop.fs.FSDataInputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.EnumSet;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Constructor;
+
+// TODO - this is the wrong package name, it should begin with org.apache
+import org.apache.parquet.Log;
+import parquet.org.apache.thrift.TBase;
+import parquet.org.apache.thrift.TException;
+import parquet.org.apache.thrift.protocol.*;
+import parquet.org.apache.thrift.transport.TIOStreamTransport;
+import parquet.org.apache.thrift.transport.TTransport;
+import parquet.org.apache.thrift.transport.TTransportException;
+
+public class CompatibilityUtil {
+  private static final boolean useV21;
+  private static final Log LOG = Log.getLog(CompatibilityUtil.class);
+  public static final V21FileAPI fileAPI;
+  private static final int MAX_SIZE = 1 << 20;
+  private static final Object bufferPool;
+  
+  private static class V21FileAPI {
+    private final Constructor<?> ELASTIC_BYTE_BUFFER_CONSTRUCTOR;
+    private final Class<?> ElasticByteBufferCls;
+    private final Class<?> ByteBufferCls;
+    private final Class<? extends Enum> ReadOptionCls;
+    private final Method READ_METHOD;
+    private final Method RELEASE_BUFFER_METHOD;
+    private final Method GET_BUFFER_METHOD;
+    private final Method PUT_BUFFER_METHOD;
+    private final Class<?> FSDataInputStreamCls;
+    
+    private V21FileAPI() throws ClassNotFoundException, NoSuchMethodException, SecurityException {
+      final String PACKAGE = "org.apache.hadoop";
+      ElasticByteBufferCls = Class.forName(PACKAGE + ".io.ElasticByteBufferPool");
+      ELASTIC_BYTE_BUFFER_CONSTRUCTOR = ElasticByteBufferCls.getConstructor();
+      ByteBufferCls = Class.forName(PACKAGE + ".io.ByteBufferPool");
+      FSDataInputStreamCls = Class.forName(PACKAGE + ".fs.FSDataInputStream");
+      ReadOptionCls = (Class<Enum>)Class.forName(PACKAGE + ".fs.ReadOption");
+      READ_METHOD = FSDataInputStreamCls.getMethod("read", ByteBufferCls, int.class, EnumSet.class);
+      RELEASE_BUFFER_METHOD = FSDataInputStreamCls.getMethod("releaseBuffer", ByteBuffer.class);
+      GET_BUFFER_METHOD = ElasticByteBufferCls.getMethod("getBuffer", boolean.class, int.class);
+      PUT_BUFFER_METHOD = ElasticByteBufferCls.getMethod("putBuffer", ByteBuffer.class);
+    }
+  }
+  
+  static {
+    boolean v21 = true;
+    try {
+      Class.forName("org.apache.hadoop.io.ElasticByteBufferPool");
+    } catch (ClassNotFoundException cnfe) {
+      v21 = false;
+    }
+    
+    useV21 = v21;
+    try {
+      if (v21) {
+        fileAPI = new V21FileAPI();
+        bufferPool = fileAPI.ELASTIC_BYTE_BUFFER_CONSTRUCTOR.newInstance();
+      } else {
+        fileAPI = null;
+        bufferPool = null;
+      }
```

Comment:
we should initialize the correct implementation of the API for all operations here and not do if (useV1) anymore anywhere in this class.